src/awkward/_v2/cpp-headers/awkward/LayoutBuilder.h

```diff
@@ -0,0 +1,1706 @@
+    _append_begin() {
+      if (current_index_ == 8) {
+        current_byte_ = uint8_t(0);
+        current_byte_ref_ = mask_.append_and_get_ref(current_byte_);
+        current_index_ = 0;
+      }
+    }
+
+    void
+    _append_end() {
+      current_index_ += 1;
+      if (valid_when_) {
+        current_byte_ref_ = current_byte_;
+      }
+      else {
+        current_byte_ref_ = ~current_byte_;
+      }
+    }
+
+    GrowableBuffer<uint8_t> mask_;
+    BUILDER content_;
+    size_t id_;
+    std::string parameters_;
+    bool valid_when_ = VALID_WHEN;
+    bool lsb_order_ = LSB_ORDER;
+    uint8_t current_byte_;
+    uint8_t& current_byte_ref_;
+    size_t current_index_;
+    uint8_t cast_[8];
+  };
+
+  template <unsigned SIZE, typename BUILDER>
+  class Regular {
+  public:
+    Regular()
+        : length_(0) {
+      size_t id = 0;
+      set_id(id);
+    }
+
+    size_t
+    length() const noexcept {
+      return length_;
+    }
+
+    void
+    clear() noexcept {
+      content_.clear();
+    }
+
+    BUILDER&
+    content() {
+      return content_;
+    }
+
+    void
+    set_id(size_t &id) noexcept {
+      id_ = id;
+      id++;
+      content_.set_id(id);
+    }
+
+    std::string parameters() const noexcept {
+      return parameters_;
+    }
+
+    void
+    set_parameters(std::string parameter) noexcept {
+      parameters_ = parameter;
+    }
+
+    bool
+    is_valid(std::string& error) const noexcept {
+      if (content_.length() != length_ * size_) {
+        std::stringstream out;
+        out << "Regular node" << id_ << "has content length " << content_.length()
+            << ", but length " << length_ << " and size " << size_;
+        error.append(out.str());
+
+        return false;
+      }
+      else {
+        return content_.is_valid(error);
+      }
+    }
+
+    BUILDER&
+    begin_list() noexcept {
+      return content_;
+    }
+
+    void
+    end_list() noexcept {
+      length_++;
+    }
+
+    void
+    buffer_nbytes(std::map<std::string, size_t> &names_nbytes) const noexcept {
+      content_.buffer_nbytes(names_nbytes);
+    }
+
+    void
+    to_buffers(std::map<std::string, void*> &buffers) const noexcept {
+      content_.to_buffers(buffers);
+    }
+
+    std::string
+    form() const noexcept {
+      std::stringstream form_key;
+      form_key << "node" << id_;
+      std::string params("");
+      if (parameters_ == "") { }
+      else {
+        params = std::string(", \"parameters\": " + parameters_);
+      }
+      return "{ \"class\": \"RegularArray\", \"content\": " + content_.form()
+                + ", \"size\": " + std::to_string(size_)  + params
+                + ", \"form_key\": \"" + form_key.str() + "\" }";
+    }
+
+  private:
+    BUILDER content_;
+    size_t id_;
+    size_t length_;
+    std::string parameters_;
+    size_t size_ = SIZE;
+  };
+
+  // FIXME: template index: int64_t and int32_t
+  template <unsigned INITIAL,  typename... BUILDERS>
+  class Union {
```

Comment:
I think, `Union` should take an additional index type parameter to allow its `GrowableBuffer<int32_t> index_`

Reply:
Union should be parameterized by a TAG type, though that has always been `uint8`, and an INDEX type, `int32`, `uin32`, or `int64`.

See v1 UnionArray: it's specialized for `UnionArray8_32`, `UnionArray8_U32`, and `UnionArray8_64`.

(Actually, I think that RNTuple unions have equal-sized TAG and INDEX types, and both of them are 32-bit. But they're also interleaved, and for an Index, we'd have to copy to make it contiguous, anyway, so the motivation to match RNTuple's TAG and INDEX types is gone.)